ambry-commons/src/main/java/com/github/ambry/commons/PipedAsyncWritableChannel.java

```diff
@@ -41,22 +49,33 @@
 public class PipedAsyncWritableChannel implements AsyncWritableChannel {
 
   private final ReadableStreamChannel sourceChannel;
+  private final int secondaryTimeoutInMs;
   private final PipedReadableStreamChannel pipedPrimaryReadChannel;
   private PipedReadableStreamChannel pipedSecondaryReadChannel = null;
   private final ReentrantLock lock = new ReentrantLock();
   private final AtomicBoolean channelOpen = new AtomicBoolean(true);
   private static final Logger logger = LoggerFactory.getLogger(PipedAsyncWritableChannel.class);
+  final static HashedWheelTimer wheel =
+      new HashedWheelTimer(new Utils.SchedulerThreadFactory("secondary-read-channel-timeout-handler-", false), 10,
+          TimeUnit.MILLISECONDS, 1024);
+  private final Metrics metrics;
 
   /**
-   * @param sourceChannel The channel that contains the stream of bytes to be read into primary and secondary destinations
-   * @param withSecondary if {@code true}, sends the bytes from source channel to secondary destination as well
+   * @param sourceChannel        The channel that contains the stream of bytes to be read into primary and secondary
+   *                             destinations
+   * @param withSecondary        if {@code true}, sends the bytes from source channel to secondary destination as well
+   * @param secondaryTimeoutInMs time in milliseconds after which secondary read channel will be closed
+   * @param metricRegistry
    */
-  public PipedAsyncWritableChannel(ReadableStreamChannel sourceChannel, boolean withSecondary) {
+  public PipedAsyncWritableChannel(ReadableStreamChannel sourceChannel, boolean withSecondary, int secondaryTimeoutInMs,
+      MetricRegistry metricRegistry) {
     this.sourceChannel = sourceChannel;
+    this.secondaryTimeoutInMs = secondaryTimeoutInMs;
 
     pipedPrimaryReadChannel = new PipedReadableStreamChannel(true);
     if (withSecondary) {
       pipedSecondaryReadChannel = new PipedReadableStreamChannel(false);
+      wheel.start();
```

Comment:
wheel.start should be threadsafe and idempotent so we can call start here in every pipe. but a better way is to have a static statement right after the variable to start it. It's probably fine but would be nice to fine some place it close it.

Reply:
Sure.. I will try to update it when I move this class to closed source..